contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,519 @@
+        SendToHub,
+        ReceiveFromHub
+    }
+
+    struct InterchainTransfer {
+        uint256 messageType;
+        bytes32 tokenId;
+        bytes sourceAddress;
+        bytes destinationAddress;
+        uint256 amount;
+        bytes data;
+    }
+
+    struct DeployInterchainToken {
+        uint256 messageType;
+        bytes32 tokenId;
+        string name;
+        string symbol;
+        uint8 decimals;
+        bytes minter;
+    }
+
+    struct SendToHub {
+        uint256 messageType;
+        string destination_chain;
+        bytes message;
+    }
+
+    struct ReceiveFromHub {
+        uint256 messageType;
+        string source_chain;
+        bytes message;
+    }
+}
+
+impl Message {
+    pub fn abi_encode(self, env: &Env) -> Bytes {
+        let msg = match self {
+            Self::InterchainTransfer(types::InterchainTransfer {
+                token_id,
+                source_address,
+                destination_address,
+                amount,
+                data,
+            }) => InterchainTransfer {
+                messageType: U256::from(types::MessageType::InterchainTransfer as u32),
+                tokenId: FixedBytes::<32>::new(token_id.into()),
+                sourceAddress: source_address.to_alloc_vec().into(),
+                destinationAddress: destination_address.to_alloc_vec().into(),
+                amount: amount.try_into().unwrap_or_default(),
+                data: data.map(|d| d.to_alloc_vec()).unwrap_or_default().into(),
+            }
+            .abi_encode_params(),
+            Self::DeployInterchainToken(types::DeployInterchainToken {
+                token_id,
+                name,
+                symbol,
+                decimals,
+                minter,
+            }) => DeployInterchainToken {
+                messageType: U256::from(types::MessageType::DeployInterchainToken as u32),
+                tokenId: FixedBytes::<32>::new(token_id.into()),
+                name: soroban_string_to_string(name),
+                symbol: soroban_string_to_string(symbol),
+                decimals: decimals.try_into().unwrap_or_default(),
+                minter: minter.map(|m| m.to_alloc_vec()).unwrap_or_default().into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
+    }
+
+    pub fn abi_decode(env: &Env, payload: &Bytes) -> Result<Self, MessageError> {
+        ensure!(payload.len() >= 32, MessageError::InsufficientMessageLength);
+
+        let binding = payload.to_alloc_vec();
+        let payload_array = binding.as_slice();
+
+        let message_type = match MessageType::abi_decode(&payload_array[0..32], true) {
+            Ok(value) => value,
+            Err(_) => return Err(MessageError::InvalidMessageType),
+        };
+
+        let message = match message_type {
+            MessageType::InterchainTransfer => {
+                let decoded = match InterchainTransfer::abi_decode_params(payload_array, true) {
```

Comment:
Could this be written as a one-liner?
```suggestion
               let decoded = InterchainTransfer::abi_decode_params(payload_array, true).map_err(|_| MessageError::AbiDecodeFailed)?;
```